File: v3-sandbox/prototype-java/src/main/java/org/hiero/common/HbarUnit.java
package org.hiero.common;

import java.util.List;
import org.jspecify.annotations.NonNull;

/**
 * The different units of Hbar. Mirrors the {@code common.HbarUnit} enum from the
 * meta-language definition.
 *
 * <p>Both attributes ({@code symbol}, {@code tinybars}) are declared {@code @@immutable}
 * in the meta-language and therefore become {@code final} fields set via the constructor.
 */
public enum HbarUnit {

    TINYBAR("tℏ", 1L),
    MICROBAR("μℏ", 100L),
    MILLIBAR("mℏ", 100_000L),
    HBAR("ℏ", 100_000_000L),
    KILOBAR("kℏ", 1_000L * 100_000_000L),
    MEGABAR("Mℏ", 1_000_000L * 100_000_000L),
    GIGABAR("Gℏ", 1_000_000_000L * 100_000_000L);

    private final String symbol;

    private final long tinybars;

    HbarUnit(@NonNull final String symbol, final long tinybars) {
        // No Objects.requireNonNull here on purpose: enum literals are compile-time
        // constants and the constructor is private — null is not reachable.
        this.symbol = symbol;
        this.tinybars = tinybars;
    }

    /**
     * Returns the human readable symbol of this unit (e.g. {@code "ℏ"}).
     *
     * @return the unit symbol
     */
    @NonNull
    public String getSymbol() {
        return symbol;
    }

    /**
     * Returns the number of tinybars contained in one unit of this kind.
     *
     * @return the number of tinybars in one unit
     */
    public long getTinybars() {
        return tinybars;
    }

    /**
     * Returns all defined {@link HbarUnit} values as an immutable list. This is the
     * Java mapping of the meta-language {@code static list<HbarUnit> values()} factory.
     *
     * @return the list of all units, never {@code null}
     */
    @NonNull
    public static List<HbarUnit> valuesAsList() {
        return List.of(values());
    }
}

File: v3-sandbox/prototype-typescript/src/common/HbarUnit.ts
export class HbarUnit {
  public static readonly TINYBAR  = new HbarUnit("tℏ", 1n);
  public static readonly MICROBAR = new HbarUnit("μℏ", 100n);
  public static readonly MILLIBAR = new HbarUnit("mℏ", 100_000n);
  public static readonly HBAR     = new HbarUnit("ℏ",  100_000_000n);
  public static readonly KILOBAR  = new HbarUnit("kℏ", 100_000_000_000n);
  public static readonly MEGABAR  = new HbarUnit("Mℏ", 100_000_000_000_000n);
  public static readonly GIGABAR  = new HbarUnit("Gℏ", 100_000_000_000_000_000n);

  private static readonly _all: HbarUnit[] = [
    HbarUnit.TINYBAR, HbarUnit.MICROBAR, HbarUnit.MILLIBAR,
    HbarUnit.HBAR, HbarUnit.KILOBAR, HbarUnit.MEGABAR, HbarUnit.GIGABAR,
  ];

  public readonly symbol: string;
  public readonly tinybars: bigint;

  private constructor(symbol: string, tinybars: bigint) {
    this.symbol = symbol;
    this.tinybars = tinybars;
  }

  public static values(): HbarUnit[] {
    return [...HbarUnit._all];
  }
}
